presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,548 @@
+import static com.facebook.presto.operator.aggregation.histogram.HashUtil.nextProbeLinear;
+import static com.facebook.presto.operator.aggregation.histogram.HashUtil.nextSumOfCount;
+import static com.facebook.presto.operator.aggregation.histogram.HashUtil.nextSumOfSquares;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INSUFFICIENT_RESOURCES;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static it.unimi.dsi.fastutil.HashCommon.arraySize;
+import static it.unimi.dsi.fastutil.HashCommon.murmurHash3;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * implementation that uses groupId in the hash key, so that we may store all groupId x value -> count in one giant map. The value, however,
+ * is normalized into a single ValueBuilder. Further, we do not have per-group instances of objects. In order to construct a histogram for a group on
+ * demand, a linked list for each group is stored that overlays all virtual nodes (a shared index, i, across parallel arrays).
+ * <p>
+ * <pre>
+ * eg,
+ * heads[100] -> 3
+ *
+ * means the first piece of data is (values[valuePositions[3]], counts[3])
+ * If next[3] is 10, then we look at (values[valuePositions[10]], counts[10])
+ *
+ * and so on to construct the value, count pairs needed. An iterator-style function, readAllValues, exists and the caller may do whatever want with the
+ * values.
+ *
+ * </pre>
+ */
+
+public class GroupedTypedHistogram
+        implements TypedHistogram
+{
+    static final float MAX_FILL_RATIO = 0.5f;
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedTypedHistogram.class).instanceSize();
+    private static final int EMPTY_BUCKET = -1;
+    private static final int NULL = -1;
+    // JMH work indicated this was superior to linear or sum_of_count
+    private static final ProbeType PROBE_TYPE = ProbeType.SUM_OF_SQUARE;
+    private final int bucketId;
+
+    private final Type type;
+    private final BlockBuilder values;
+    private final BucketNodeFactory bucketNodeFactory;
+    //** these parallel arrays represent a node in the hash table
+    private final LongBigArray counts;
+    // need to store the groupId for a node for when we are doing value comparisons in hash lookups
+    private final LongBigArray groupIds;
+    // array of nodePointers (index in counts, valuePositions)
+    private final IntBigArray nextPointers;
+    // since we store histogram values in a hash, two histograms may have the same position (each unique value should have only one position in the internal
+    // BlockBuilder of values; not that we extract a subset of this when constructing per-group-id histograms)
+    private final IntBigArray valuePositions;
+    private final LongBigArray valueHashList = new LongBigArray();
```

Comment:
I can init in the constructor like the rest.
but it's different--for a given node, it's the hash of hte value, pre-computed

it saved time looking up values in ValueStore

            int nextValuePosition = valueStore.getPosition(type, block, position, valueHash);

we use it. again, via profiling, I found the space/cpu tradeoff helped when we saw the same value > 1

(since it's mostly an array, and mostly constant save the internal array boundaries, it saved time)
now, are they same?

outside one: bucketId -> valueHash (no groupId, no mask)

inside ValueStore -> location in output Builder/Block and the valueHash for it. when we do a rehash, we need access to this list by the buckets within ValueStore to rehash everything.

think of outside as global bucket -> valueHash
ValueStore: internal bucket -> valueHash

both save on recomputation, the outer on repeated add of the same value, the internal on rehash()

and recall outer hash is group x valueHash -> node (containing valueHash
ValueStore is just valueHash (+ value for collisions)  -> valuePointer

talk more in person--pretty sure we need both. they optimized away a lot of cpu when i did yourkit, and they are definitely distinct